internal/storage/chunk_uploader.go

```diff
@@ -0,0 +1,233 @@
+	}
+
+	if req.GenerationPrecondition != nil && *req.GenerationPrecondition != 0 {
+		return nil, fmt.Errorf("request received for pre-existing object %s, supported only for new objects", req.Name)
+	}
+
+	if chunkSize <= 0 {
+		return nil, fmt.Errorf("chunkSize <= 0")
+	}
+
+	// Raw initialization.
+	uploader := chunkUploader{}
+
+	// Store references to necessary parameters.
+	uploader.objectName = obj.ObjectName()
+	uploader.userProgressFunc = progressFunc
+
+	// Create a new writer with the requested attributes, using Go Storage Client.
+	// NewWriter never returns nil, so no nil-check is needed on it.
+	wc := obj.NewWriter(ctx)
+	wc = storageutil.SetAttrsInWriter(wc, req)
+	wc.ChunkSize = chunkSize
+	wc.ProgressFunc = func(n int64) {
+		uploader.totalBytesUploadSuccessfullySoFar.Store(n)
+		logger.Debugf("%d bytes copied so far for object/file %s. chunk-size = %d", n, req.Name, wc.ChunkSize)
+
+		if uploader.userProgressFunc != nil {
+			uploader.userProgressFunc(n)
+		}
+	}
+
+	uploader.writer = wc
+	uploader.state = Initialized
+
+	return &uploader, nil
+}
+
+// BytesUploadedSoFar returns the total number of bytes successfully uploaded
+// so far using this uploader.
+// This is thread-safe against the
+// in-progress calls to BytesUploadedSoFar/Upload/Close
+// invoked from other threads/go-routines.
+func (uploader *chunkUploader) BytesUploadedSoFar() int64 {
+	return uploader.totalBytesUploadSuccessfullySoFar.Load()
+}
+
+func (uploader *chunkUploader) readyToUpload() bool {
+	switch uploader.state {
+	case Initialized, Uploading:
+		return true
+	default:
+		return false
+	}
+}
+
+func (uploader *chunkUploader) readyToClose() bool {
+	switch uploader.state {
+	case Initialized, Uploading, UploadError:
+		return true
+	default:
+		return false
+	}
+}
+
+// Upload uploads the passed content to GCS.
+// This waits (using mutex) until the completion of
+// in-progress calls to BytesUploadedSoFar/Upload/Close
+// invoked from other threads/go-routines.
+// If contents contains more than n chunks, then it synchronizes on the
+// upload of the first n chunks, and queues up the last chunk,
+// before returning. Ideally, the caller would want to invoke this
+// function in a goroutine in case of contents containing multiple chunks.
+func (uploader *chunkUploader) Upload(ctx context.Context, contents io.Reader) error {
+	if ctx == nil {
+		return fmt.Errorf("ctx is nil")
+	}
+
+	uploader.mu.Lock()
+	defer uploader.mu.Unlock()
+
+	if !uploader.readyToUpload() {
+		return fmt.Errorf("uploader not ready to upload: object: %s, status = %v, writer: %v", uploader.objectName, uploader.state, uploader.writer)
+	}
+
+	var n int64
+	var err error
+	ioCopyComplete := make(chan bool)
+	go func() {
+		n, err = io.Copy(uploader.writer, contents)
+		ioCopyComplete <- true
+	}()
+
+	select {
+	case <-ctx.Done():
+		return fmt.Errorf("uploader.Upload() timed out")
```

Comment:
nit: whether it's timed out or cancelled because of other reasons, we don't know. So, can we use ctx.Err() here instead of hard coding timed out?

Reply:
Done.